src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -188,46 +198,80 @@ private void EnqueueProcessSnapshot(Document newDocument)
+            private async Task EnqueueProcessSnapshotWorkerAsync(Document currentDocument, CancellationToken cancellationToken)
             {
                 // we will enqueue new one soon, cancel pending refresh right away
                 _reportChangeCancellationSource.Cancel();
 
-                var newText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
-                var snapshot = newText.FindCorrespondingEditorTextSnapshot();
-                if (snapshot == null)
+                var currentText = await currentDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var currentSnapshot = currentText.FindCorrespondingEditorTextSnapshot();
+                if (currentSnapshot == null)
                 {
                     // It's possible that we're seeing a notification for an update that happened
                     // just before the file was opened, and so the document we're given is still the
                     // old one.
                     return;
                 }
 
-                var latencyTracker = new RequestLatencyTracker(SyntacticLspLogger.RequestType.SyntacticTagger);
-                using (latencyTracker)
+                var service = TryGetClassificationService(currentSnapshot);
+                object currentCachedData = null;
+                var changedSpan = currentSnapshot.GetFullSpan();
+                if (service != null)
                 {
-                    // preemptively parse file in background so that when we are called from tagger from UI thread, we have tree ready.
-                    // F#/typescript and other languages that doesn't support syntax tree will return null here.
-                    _ = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                    var latencyTracker = new RequestLatencyTracker(SyntacticLspLogger.RequestType.SyntacticTagger);
+                    using (latencyTracker)
+                    {
+                        // preemptively allow the classification service to compute and cache data for this file.  For
+                        // example, in C# and VB we will parse the file so that are called from tagger from UI thread,
+                        // we have the root of the tree ready to go.
+                        currentCachedData = await service.GetDataToCacheAsync(currentDocument, cancellationToken).ConfigureAwait(false);
+
+                        // Query the service to determine waht span of the document actually changed and should be
+                        // reclassified in the host editor.
+                        changedSpan = await GetChangedSpanAsync(currentDocument, currentSnapshot, service, cancellationToken).ConfigureAwait(false);
+                    }
                 }
 
+                // Once we're past this point, we're mutating our internal state so we cannot cancel past that.
+                cancellationToken = CancellationToken.None;
```

Comment:
It's kind of odd that we're setting the cancellationToken like this. Can we separate everything that mutates internal state into a separate method to avoid having to do this?

Reply:
We can.  However, that opens up the possibility for mistakes.  For example, if that code accidentally passes in a CT in the future, or gets one ambiently.  I've found it works really well to keep the patterns all identical (i.e. async method get passed the contextual CT) but to just assign that CT to an uncancellable form.  THis keeps parties not having to coordinate on anything, and helps prevent future accidents.

Reply:
Maybe, my gut is that it seems too easy to accidently put stuff past that point and not know it's not the same cancellation token. It's just a hunch though and no real data. Not worth blocking

Reply:
i removed it.